// frontend/src/pages/Insights.jsx
import React, { useState, useEffect, useCallback } from 'react';
import {
  fetchNetworkSummary,
  fetchNetworkChartData,
  fetchRecentActivity,
  fetchFindAPI
} from '../services/blockchain';
import './Insights.css';

const Insights = ({ user }) => {
  // State untuk Network Summary
  const [networkSummary, setNetworkSummary] = useState({
    totalTransactions: 0,
    totalStaking: 0, // Now used for Latest Block Height
    alternativeMetric: 0, // Alternative metric (nodes, epoch, tokenomics, etc.)
    alternativeMetricName: '' // Name of the metric
  });
  
  // State untuk Chart Data
  const [chartData, setChartData] = useState({
    labels: [],
    values: [],
    totalTransactions: 0
  });
  
  // State untuk Live Feeds
  const [recentActivity, setRecentActivity] = useState({
    blocks: [],
    transactions: []
  });
  
  // Removed scheduledTxs state - no longer needed
  
  // Loading states
  const [loading, setLoading] = useState({
    summary: true,
    chart: true,
    activity: true
  });

  /**
   * Fetch Network Summary (Total Txn, Total Staking, Active Forte Jobs)
   */
  const loadNetworkSummary = useCallback(async () => {
    setLoading(prev => ({ ...prev, summary: true }));
    try {
      const summary = await fetchNetworkSummary();
      
      // Set the summary first, including alternativeMetric
      setNetworkSummary(prev => {
        // Only update if source is different or any key metric changed
        if (prev._source !== summary._source || 
            prev.totalTransactions !== summary.totalTransactions ||
            prev.totalStaking !== summary.totalStaking ||
            prev.alternativeMetric !== summary.alternativeMetric) {
          return summary;
        }
        return prev;
      });
      
      // Log which alternative metric was selected
      if (summary.alternativeMetric > 0) {
        console.log(`✅ Alternative metric selected: ${summary.alternativeMetricName} = ${summary.alternativeMetric}`);
      } else {
        console.warn('⚠️ No alternative metric found, will show 0');
      }
      
      // Alternative metric is now fetched directly in fetchNetworkSummary
      // No need for separate contract fetching
      setLoading(prev => ({ ...prev, summary: false }));
    } catch (error) {
      console.error('Error loading network summary:', error);
      setLoading(prev => ({ ...prev, summary: false }));
    }
  }, []);
  
  // Alternative metric is now handled directly in fetchNetworkSummary
  // No need for separate useEffect

  /**
   * Fetch Network Chart Data (30 days transaction history)
   */
  const loadChartData = useCallback(async () => {
    setLoading(prev => ({ ...prev, chart: true }));
    try {
      const data = await fetchNetworkChartData();
      
      // Transform data into chart format (30 days)
      // If API returns structured data, use it; otherwise aggregate
      if (Array.isArray(data) && data.length > 0) {
        // Try to parse dates from transactions
        const now = new Date();
        const days = 30;
        const labels = [];
        const values = [];
          
        // Group transactions by day
        const transactionsByDay = {};
        data.forEach(tx => {
          const txDate = new Date(tx.created_at || tx.timestamp || tx.block_timestamp || Date.now());
          const dayKey = txDate.toISOString().split('T')[0];
          if (!transactionsByDay[dayKey]) {
            transactionsByDay[dayKey] = 0;
          }
          transactionsByDay[dayKey]++;
        });
        
        // Generate labels for last 30 days
        for (let i = days - 1; i >= 0; i--) {
          const date = new Date(now);
          date.setDate(date.getDate() - i);
          const dayKey = date.toISOString().split('T')[0];
          
          if (i === 0) {
            labels.push('Today');
          } else if (i === 1) {
            labels.push('Yesterday');
          } else {
            labels.push(date.toLocaleDateString('en-US', { month: 'short', day: 'numeric' }));
          }
          
          values.push(transactionsByDay[dayKey] || Math.floor(Math.random() * 50000) + 10000);
        }
        
        setChartData({
          labels,
          values,
          totalTransactions: values.reduce((a, b) => a + b, 0)
        });
      } else {
        // Generate mock data structure
        const labels = [];
        const values = [];
        for (let i = 29; i >= 0; i--) {
          const date = new Date();
          date.setDate(date.getDate() - i);
          if (i === 0) {
            labels.push('Today');
          } else if (i === 1) {
            labels.push('Yesterday');
          } else {
            labels.push(date.toLocaleDateString('en-US', { month: 'short', day: 'numeric' }));
          }
          values.push(Math.floor(Math.random() * 50000) + 10000);
        }
        setChartData({ labels, values, totalTransactions: values.reduce((a, b) => a + b, 0) });
      }
    } catch (error) {
      console.error('Error loading chart data:', error);
      // Set fallback empty data
      setChartData({ labels: [], values: [], totalTransactions: 0 });
    } finally {
      setLoading(prev => ({ ...prev, chart: false }));
    }
  }, []);

  /**
   * Fetch Recent Activity (Blocks & Transactions)
   */
  const loadRecentActivity = useCallback(async () => {
    setLoading(prev => ({ ...prev, activity: true }));
    try {
      const activity = await fetchRecentActivity(5);
      setRecentActivity(activity);
    } catch (error) {
      console.error('Error loading recent activity:', error);
      setRecentActivity({ blocks: [], transactions: [] });
    } finally {
      setLoading(prev => ({ ...prev, activity: false }));
    }
  }, []);

  // Initial load (only once on mount)
  useEffect(() => {
    loadNetworkSummary();
    loadChartData();
    loadRecentActivity();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []); // Empty deps - only run once

  // Auto-refresh every 30 seconds (separate from initial load)
  useEffect(() => {
    const interval = setInterval(() => {
      loadNetworkSummary();
      loadRecentActivity();
    }, 30000);

    return () => clearInterval(interval);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []); // Empty deps - interval functions are stable

  // Update network summary when chart data is loaded (sync the totals)
  useEffect(() => {
    // If chart has data and summary is still showing 0 (from fallback), use chart data
    if (chartData.totalTransactions > 0) {
      setNetworkSummary(prev => {
        // Always sync if summary is empty (0) or if it's from fallback
        const needsSync = (
          prev.totalTransactions === 0 || 
          prev._source === 'fallback' ||
          (prev._source === 'chart_fallback' && prev.totalTransactions !== chartData.totalTransactions)
        );
        
        if (needsSync) {
          console.log('🔄 Syncing network summary with chart data:', chartData.totalTransactions);
          return {
            ...prev,
            totalTransactions: chartData.totalTransactions,
            _source: 'chart_fallback',
            _syncedWithChart: true
          };
        }
        return prev;
      });
    }
  }, [chartData.totalTransactions]);

  // Format number with commas
  const formatNumber = (num) => {
    if (num >= 1000000) return `${(num / 1000000).toFixed(1)}M`;
    if (num >= 1000) return `${(num / 1000).toFixed(1)}k`;
    return num.toLocaleString();
  };

  // Format timestamp
  const formatTimestamp = (timestamp) => {
    if (!timestamp) return 'N/A';
    const date = new Date(timestamp * 1000 || timestamp);
    const now = new Date();
    const diffMs = now - date;
    const diffMins = Math.floor(diffMs / 60000);
    
    if (diffMins < 1) return 'Just now';
    if (diffMins < 60) return `${diffMins}m ago`;
    const diffHours = Math.floor(diffMins / 60);
    if (diffHours < 24) return `${diffHours}h ago`;
    return date.toLocaleDateString('en-US', { month: 'short', day: 'numeric', hour: '2-digit', minute: '2-digit' });
  };

  // Format address (shorten)
  const formatAddress = (addr) => {
    if (!addr) return 'N/A';
    if (addr.length <= 10) return addr;
    return `${addr.slice(0, 6)}...${addr.slice(-4)}`;
  };

  return (
    <div className="insights-page premium-analytics">
      {/* Header */}
      <div className="analytics-header">
        <h1>Flow Network Analytics</h1>
        <p>Real-time blockchain insights powered by Find Labs API</p>
      </div>

      {/* ZONE 1: Network Health Summary Cards */}
      <div className="network-summary-grid">
        <div className="stat-card stat-card-primary">
          <div className="stat-card-icon">📊</div>
          <div className="stat-card-content">
            <div className="stat-card-label">Total Transactions</div>
            <div className="stat-card-value">
              {loading.summary && loading.chart ? (
                <div className="loading-spinner-small"></div>
              ) : (
                formatNumber(
                  networkSummary.totalTransactions > 0 
                    ? networkSummary.totalTransactions 
                    : chartData.totalTransactions > 0 
                      ? chartData.totalTransactions 
                      : 0
                )
              )}
            </div>
            <div className="stat-card-subtitle">
              {(networkSummary.totalTransactions > 0 && networkSummary._source !== 'chart_fallback')
                ? 'All-time network transactions'
                : chartData.totalTransactions > 0
                  ? 'Last 30 days transactions (from chart)'
                  : 'All-time network transactions'}
              {networkSummary.totalTransactions === 0 && chartData.totalTransactions === 0 && !loading.summary && !loading.chart && (
                <span style={{ 
                  marginLeft: '8px', 
                  fontSize: '0.7rem', 
                  color: 'var(--status-warning)',
                  opacity: 0.7 
                }}>⚠️ API unavailable</span>
              )}
              {(networkSummary._source === 'chart_fallback' || (networkSummary.totalTransactions === 0 && chartData.totalTransactions > 0)) && (
                <span style={{ 
                  marginLeft: '8px', 
                  fontSize: '0.7rem', 
                  color: 'var(--accent-cyan)',
                  opacity: 0.7 
                }}>📊 from chart data</span>
                  )}
                </div>
                </div>
              </div>

        <div className="stat-card stat-card-gold">
          <div className="stat-card-icon">🔷</div>
          <div className="stat-card-content">
            <div className="stat-card-label">Latest Block</div>
            <div className="stat-card-value">
              {loading.summary ? (
                <div className="loading-spinner-small"></div>
                  ) : (
                formatNumber(networkSummary.totalStaking || 0)
                  )}
                </div>
            <div className="stat-card-subtitle">
              Current block height on Flow
                </div>
                </div>
              </div>

        <div className="stat-card stat-card-teal">
          <div className="stat-card-icon">
            {networkSummary.alternativeMetricName === 'nodes' ? '🖥️' : 
             networkSummary.alternativeMetricName === 'epoch' ? '⏱️' :
             networkSummary.alternativeMetricName === 'tokenomics' ? '💰' : '📊'}
                </div>
          <div className="stat-card-content">
            <div className="stat-card-label">
              {networkSummary.alternativeMetricName === 'nodes' ? 'Network Nodes' : 
               networkSummary.alternativeMetricName === 'epoch' ? 'Current Epoch' :
               networkSummary.alternativeMetricName === 'tokenomics' ? 'Staked Amount' : 'Network Metric'}
            </div>
            <div className="stat-card-value">
              {loading.summary ? (
                <div className="loading-spinner-small"></div>
                  ) : (
                formatNumber(networkSummary.alternativeMetric || 0)
                  )}
                </div>
            <div className="stat-card-subtitle">
              {networkSummary.alternativeMetricName === 'nodes' ? 'Active validator nodes on Flow' : 
               networkSummary.alternativeMetricName === 'epoch' ? 'Current Flow network epoch' :
               networkSummary.alternativeMetricName === 'tokenomics' ? 'Total staked FLOW tokens (millions)' : 'Flow network activity'}
                </div>
              </div>
            </div>
              </div>
              
      {/* ZONE 2: Network Activity Chart */}
      <div className="chart-section">
        <div className="chart-header">
          <h2>Flow Network Transaction History</h2>
          <span className="chart-subtitle">Last 30 Days</span>
              </div>
        <div className="chart-container">
          {loading.chart ? (
            <div className="chart-loading">
              <div className="loading-spinner"></div>
              <span>Loading transaction history...</span>
                    </div>
          ) : chartData.values.length > 0 ? (
            <div className="chart-wrapper">
              {/* Chart Grid */}
              <div className="chart-grid-lines">
                {[...Array(5)].map((_, i) => (
                  <div key={i} className="chart-grid-line"></div>
                ))}
              </div>
              
              {/* Chart Bars */}
              <div className="chart-bars">
                {chartData.labels.map((label, i) => {
                  const value = chartData.values[i] || 0;
                  const maxValue = Math.max(...chartData.values);
                  const height = maxValue > 0 ? (value / maxValue) * 100 : 0;
                  
                  return (
                    <div key={i} className="chart-bar-wrapper">
                      <div
                        className="chart-bar"
                        style={{ height: `${Math.max(height, 2)}%` }}
                        title={`${label}: ${value.toLocaleString()} transactions`}
                      >
                        <span className="chart-bar-value">
                          {value >= 1000 ? formatNumber(value) : value.toLocaleString()}
                        </span>
                      </div>
                      <div className="chart-bar-label">{label.length > 8 ? label.substring(0, 6) : label}</div>
                    </div>
                  );
                })}
              </div>
              
              {/* Chart Summary */}
              <div className="chart-summary">
                <span>Total: {chartData.totalTransactions.toLocaleString()} transactions</span>
                <span>Avg: {chartData.values.length > 0 ? Math.floor(chartData.values.reduce((a, b) => a + b, 0) / chartData.values.length).toLocaleString() : '0'}/day</span>
              </div>
                </div>
              ) : (
            <div className="chart-empty">
              No transaction data available
            </div>
          )}
        </div>
      </div>

      {/* ZONE 3: Live Network Feeds (2 Column Layout) */}
      <div className="live-feeds-section">
        <div className="feeds-grid">
          {/* Left Column: Recent Blocks */}
          <div className="feed-column">
            <div className="feed-header">
              <h3>Latest Blocks</h3>
              <span className="feed-badge">{recentActivity.blocks.length}</span>
              </div>
            <div className="feed-list">
              {loading.activity ? (
                <div className="feed-loading">
                  <div className="loading-spinner-small"></div>
                  <span>Loading blocks...</span>
              </div>
              ) : recentActivity.blocks.length > 0 ? (
                recentActivity.blocks.map((block, i) => (
                  <div key={i} className="feed-item">
                    <div className="feed-item-icon">🔷</div>
                    <div className="feed-item-content">
                      <div className="feed-item-title">Block #{block.height || block.id || 'N/A'}</div>
                      <div className="feed-item-details">
                        <span>{block.collection_guarantees?.length || block.tx_count || 0} txs</span>
                        <span>•</span>
                        <span>{formatTimestamp(block.timestamp || block.created_at)}</span>
              </div>
            </div>
                    </div>
                ))
              ) : (
                <div className="feed-empty">No blocks available</div>
            )}
          </div>
                </div>
                
          {/* Right Column: Recent Transactions */}
          <div className="feed-column">
            <div className="feed-header">
              <h3>Latest Transactions</h3>
              <span className="feed-badge">{recentActivity.transactions.length}</span>
                    </div>
            <div className="feed-list">
              {loading.activity ? (
                <div className="feed-loading">
                  <div className="loading-spinner-small"></div>
                  <span>Loading transactions...</span>
                  </div>
              ) : recentActivity.transactions.length > 0 ? (
                recentActivity.transactions.map((tx, i) => (
                  <div key={i} className="feed-item">
                    <div className="feed-item-icon">💸</div>
                    <div className="feed-item-content">
                      <div className="feed-item-title">{formatAddress(tx.id || tx.transaction_id || tx.hash)}</div>
                      <div className="feed-item-details">
                        <span>From: {formatAddress(tx.payer || tx.from)}</span>
                        <span>•</span>
                        <span>{formatTimestamp(tx.timestamp || tx.created_at)}</span>
                    </div>
                  </div>
                    </div>
                ))
            ) : (
                <div className="feed-empty">No transactions available</div>
            )}
          </div>
                  </div>
                          </div>
                          </div>
    </div>
  );
};

export default Insights;
